jupyterhub/kubespawner
# Volumes

Summary
----------------
Supports dynamically provisioning pvc and pv with storage class for single users pods. Users can now have persistent storage on their single user notebooks.

How To Test
----------------
1. Add a StorageClass to the manifest.yaml, maybe one named eg. `gce-standard-storage`
2. Configure the following variables in jupyterhub_config.py, for example:
```
c.KubeSpawner.pvc_name_template = 'claim-{username}-{userid}'
c.KubeSpawner.storage_class = 'gce-standard-storage'
c.KubeSpawner.access_modes = ['ReadWriteOnce']
c.KubeSpawner.storage = '10Gi'

c.KubeSpawner.volumes = [
    {
        'name': 'volume-{username}-{userid}',
        'persistentVolumeClaim': {
            'claimName': 'claim-{username}-{userid}'
        }
    }
]
c.KubeSpawner.volume_mounts = [
    {
        'mountPath': '/home',
        'name': 'volume-{username}-{userid}'
    }
]
```
3. Build the new jupyterhub image and apply the new manifest.
4. Log into jupyterhub as an arbitrary user, say `peter`
5. Start `peter`'s single user server
6. Create a file in `home`, or wherever you mount the volume
7. Go to control panel and shut down the server. Ensure that the pod is terminated by `watch kubectl get pods`
8. Start the server again and navigate to `home` to see if the file is still there.
9. Repeat this step for a different user starting at step 4.
10. Repeat steps 4-8 with yet another user but while both of the other users' pods are down, so nothing is mounted to their corresponding volume claims.
